sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -135,11 +135,14 @@ object TypeCoercion {
     case (DateType, StringType) => Some(StringType)
     case (StringType, TimestampType) => Some(StringType)
     case (TimestampType, StringType) => Some(StringType)
-    case (TimestampType, DateType) => Some(StringType)
-    case (DateType, TimestampType) => Some(StringType)
     case (StringType, NullType) => Some(StringType)
     case (NullType, StringType) => Some(StringType)
 
+    // Cast to TimestampType when we compare DateType with TimestampType
+    // i.e. TimeStamp('2017-03-01 00:00:00') eq Date('2017-03-01') = true
```

Comment:
So it means `TimeStamp('2017-03-01 00:00:01') eq Date('2017-03-01') = false`? Looks a bit weird.

At least this comparison does not get true in Hive:
```sql
hive> select cast('2017-03-01 00:00:00' as timestamp) = cast('2017-03-01' as date);
OK
false
```

Reply:
Thank you for your try.

The current implementation follows [this answer|https://issues.apache.org/jira/browse/SPARK-23549?focusedCommentId=16391351&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16391351].

[Which version of hive|https://issues.apache.org/jira/browse/SPARK-23549?focusedCommentId=16391877&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16391877] did you use?

Reply:
Hive 2.1.

Reply:
Thanks 
@dongjoon-hyun are you pointing out [this HIVE JIRA entry|https://issues.apache.org/jira/browse/HIVE-15236] in [this comment](https://issues.apache.org/jira/browse/SPARK-23549?focusedCommentId=16391877&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16391877)? Or other HIVE JIRA entries?

Reply:
Seems it's fixed in Hive 2.2.